File: weapon_management/weapon_management/weapon_management/report/ammunition_inventory/ammunition_inventory.js

frappe.query_reports["Ammunition Inventory"] = {
    "filters": [
        {
            "fieldname": "unit",
            "label": __("Unit"),
            "fieldtype": "Link",
            "options": "Unit Master",
            "reqd": 0
        },
        {
            "fieldname": "ammunition_name",
            "label": __("Ammunition Name"),
            "fieldtype": "Link",
            "options": "Ammunition Category Master",
            "reqd": 0
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 0
        },
        {
            "fieldname": "Clear Filters",
            "label": __("Clear Filters"),
            "fieldtype": "Button",
            "reqd": 0,
            "click": function() {
                // Clear the filter values
                frappe.query_report.set_filter_value("unit", "");
                frappe.query_report.set_filter_value("ammunition_name", "");
                frappe.query_report.set_filter_value("from_date", "");
                frappe.query_report.set_filter_value("to_date", "");
                // Update the report
                frappe.query_report.refresh();
            },
        }
    ]
};

File: weapon_management/weapon_management/weapon_management/report/ammunition_inventory/ammunition_inventory.py
# Copyright (c) 2023, Expedien and contributors
# For license information, please see license.txt


import frappe

def execute(filters=None):
    columns = ["Ammunition Name", "Available Ammunition"]

    data = []

    filters_dict = frappe._dict(filters)

    query = """
        SELECT 
            tacm.ammunition_category AS "Ammunition Name",
            SUM(taid.available_quantity) AS "Available Ammunition"
        FROM 
            "tabAmmunition In Details" taid 
        INNER JOIN 
            "tabUnit Master" tum ON taid.unit_location = tum.name
        INNER JOIN 
            "tabAmmunition Category Master" tacm ON taid.ammunition_category = tacm.name
    """

    where_clause = "WHERE 1=1" 

    if filters_dict.get("unit"):
        where_clause += f" AND tum.name = %(unit)s"

    if filters_dict.get("ammunition_name"):
        where_clause += f" AND tacm.name = %(ammunition_name)s"

    if filters_dict.get("from_date") and filters_dict.get("to_date"):
        where_clause += " AND taid.date_acquired BETWEEN %(from_date)s AND %(to_date)s"

    query += where_clause

    query += " GROUP BY tacm.ammunition_category;"

    result = frappe.db.sql(query, filters_dict, as_dict=True)

    for row in result:
        data_row = [row.get("Ammunition Name"), row.get("Available Ammunition")]
        data.append(data_row)

    return columns, data
